chain-extensions/pallet-assets/src/lib.rs

```diff
@@ -0,0 +1,379 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod weights;
+
+use crate::weights::WeightInfo as ChainExtensionWeightInfo;
+
+use assets_chain_extension_types::{GetOrigin, Origin, Outcome};
+use codec::Encode;
+use frame_support::traits::fungibles::InspectMetadata;
+use frame_support::traits::tokens::fungibles::approvals::Inspect;
+use pallet_assets::WeightInfo;
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, RetVal, SysConfig, UncheckedFrom,
+};
+use sp_runtime::traits::StaticLookup;
+use sp_runtime::DispatchError;
+use sp_std::marker::PhantomData;
+use sp_std::vec::Vec;
+
+enum AssetsFunc {
+    Create,
+    Transfer,
+    Mint,
+    Burn,
+    BalanceOf,
+    TotalSupply,
+    Allowance,
+    ApproveTransfer,
+    CancelApproval,
+    TransferApproved,
+    SetMetadata,
+    MetadataName,
+    MetadataSymbol,
+    MetadataDecimals,
+}
+
+impl TryFrom<u16> for AssetsFunc {
+    type Error = DispatchError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(AssetsFunc::Create),
+            2 => Ok(AssetsFunc::Transfer),
+            3 => Ok(AssetsFunc::Mint),
+            4 => Ok(AssetsFunc::Burn),
+            5 => Ok(AssetsFunc::BalanceOf),
+            6 => Ok(AssetsFunc::TotalSupply),
+            7 => Ok(AssetsFunc::Allowance),
+            8 => Ok(AssetsFunc::ApproveTransfer),
+            9 => Ok(AssetsFunc::CancelApproval),
+            10 => Ok(AssetsFunc::TransferApproved),
+            11 => Ok(AssetsFunc::SetMetadata),
+            12 => Ok(AssetsFunc::MetadataName),
+            13 => Ok(AssetsFunc::MetadataSymbol),
+            14 => Ok(AssetsFunc::MetadataDecimals),
+            _ => Err(DispatchError::Other(
+                "PalletAssetsExtension: Unimplemented func_id",
+            )),
+        }
+    }
+}
+
+/// Pallet Assets chain extension.
+pub struct AssetsExtension<T>(PhantomData<T>);
```

Comment:
I believe weight info should be generic type parameter that implements your defined ChainExtensionWeightInfo trait.

because chain extensions have no Config trait implementation (as it done for pallets). To push custom weight implementations we should use type magic for CE structure.

in your case it will be `AssetExtension<T, W>`, and as usual we can define `()` implementation for this trait for making it easy to use by default.